libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,400 @@
+             __is_callable_from<_Func>)
+  _LIBCPP_HIDE_FROM_ABI move_only_function(_Func&& __func) noexcept {
+    using _FuncWraps = __function_wrappers<_Func>;
+    using _UnRefFunc = remove_reference_t<_Func>;
+
+    if constexpr ((is_pointer_v<_UnRefFunc> && is_function_v<remove_pointer_t<_UnRefFunc>>) ||
+                  is_member_function_pointer_v<_UnRefFunc>) {
+      if (__func == nullptr) {
+        __storage_ = {};
+        return;
+      } else {
+        auto& __data   = __storage_.__trivially_destructible_;
+        __data.__call_ = std::bit_cast<void*>(&_FuncWraps::__call);
+        std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__data_), std::forward<_Func>(__func));
+        __data.__status_ = _Store::_Status::_TriviallyDestructible;
+      }
+    } else if constexpr (__is_move_only_function<remove_cvref_t<_Func>>::value) {
+      if (!__func) {
+        __storage_ = {};
+      } else {
+        auto& __data      = __storage_.__large_;
+        __data.__call_    = std::bit_cast<void*>(&_FuncWraps::__call);
+        __data.__destroy_ = &_FuncWraps::__destroy;
+        __data.__heap_    = static_cast<std::byte*>(::operator new(sizeof(_Func), std::align_val_t(alignof(_Func))));
+        std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__heap_), std::forward<_Func>(__func));
+        __data.__status_ = _Store::_Status::_Heap;
+      }
+    } else {
+      __construct<_Func>(std::forward<_Func>(__func));
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI _Store::_Status __get_status() noexcept { return __storage_.__large_.__status_; }
+
+  template <class _Func, class... _Args>
+    requires is_constructible_v<decay_t<_Func>, _Args...> && __is_callable_from<_Func>
+  _LIBCPP_HIDE_FROM_ABI explicit move_only_function(in_place_type_t<_Func>, _Args&&... __args) {
+    static_assert(is_same_v<decay_t<_Func>, _Func>);
+    __construct<_Func>(std::forward<_Args>(__args)...);
+  }
+
+  template <class _Func, class _InitListType, class... _Args>
+    requires is_constructible_v<decay_t<_Func>, initializer_list<_InitListType>, _Args...> && __is_callable_from<_Func>
+  _LIBCPP_HIDE_FROM_ABI explicit move_only_function(
+      in_place_type_t<_Func>, initializer_list<_InitListType> __il, _Args&&... __args) {
+    __construct<_Func>(__il, std::forward<_Args>(__args)...);
+  }
+
+  // TODO: Do we want to make this `noexcept` as an extensions?
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(move_only_function&& __other) noexcept {
+    move_only_function(std::move(__other)).swap(*this);
+    return *this;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(nullptr_t) noexcept {
+    __reset();
+    return *this;
+  }
+
+  // TODO: Do we want to make this `noexcept` as an extensions?
+  template <class _Func>
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(_Func&& __func) noexcept {
+    move_only_function(std::forward<_Func>(__func)).swap(*this);
+  }
```

Comment:
"Effects: Equivalent to:" means that we need to constrain this; otherwise `mof = 1;` is well-formed, and we certainly don't want that.

This certainly cannot be `noexcept`.
```suggestion
  template <class _Func>
    requires (!is_same_v<remove_cvref_t<_Func>, move_only_function> && 
              !__is_inplace_type<_Func>::value &&
              __is_callable_from<_Func>)
  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(_Func&& __func) {
    move_only_function(std::forward<_Func>(__func)).swap(*this);
  }
```

Reply:
Good catch with the `noexcept`! Took me a few seconds to see the bug.